ethereum/sidecar/server.go

```diff
@@ -951,11 +965,118 @@ func (s *Server) fetchAssetsUnlockConfirmedEvents(
 	return result, nil
 }
 
-//nolint:unparam
-func (s *Server) attestAssetsUnlockedEvents(ctx context.Context) error {
-	// TODO: Implement
-	<-ctx.Done()
-	return nil
+func (s *Server) unqueueAttestation() *bridgetypes.AssetsUnlockedEvent {
+	s.attestationMutex.Lock()
+	defer s.attestationMutex.Unlock()
+	if len(s.attestationQueue) == 0 {
+		return nil
+	}
+
+	attestation := s.attestationQueue[0]
+	s.attestationQueue = s.attestationQueue[1:]
+
+	return &attestation
+}
+
+func (s *Server) attestAssetsUnlockedEvents(ctx context.Context) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			for attestation := s.unqueueAttestation(); attestation != nil; attestation = s.unqueueAttestation() {
+				bridgeAssetsUnlocked := &portal.MezoBridgeAssetsUnlocked{
+					UnlockSequenceNumber: attestation.UnlockSequence.BigInt(),
+					Recipient:            attestation.Recipient,
+					Token:                common.HexToAddress(attestation.Token),
+					Amount:               attestation.Amount.BigInt(),
+					Chain:                uint8(attestation.Chain), //nolint:gosec // G115: Chain is known to be within uint8 range
+				}
+
+				ok, err := s.attestationValidator.IsValid(ctx, bridgeAssetsUnlocked)
+				if err != nil {
+					// this is only context cancellation
+					return
+				}
+				if !ok {
+					continue
+				}
+
+				ok, err = s.attestationValidator.IsConfirmed(bridgeAssetsUnlocked)
+				if err != nil {
+					s.logger.Error("couldn't confirm attestation", "attestation", attestation.String(), "error", err)
```

Comment:
We are just logging because we want to fall into the inner loop and rety IsConfirmed, right? If so, let's drop a comment explaining this. Let's also modify the log error to say we are continuing with attestations.

Reply:
2a245746